searchlib/src/vespa/searchlib/expression/and_predicate_node.cpp

```diff
@@ -0,0 +1,51 @@
+// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#include "and_predicate_node.h"
+#include "resultnode.h"
+
+#include <vespa/vespalib/objects/serializer.hpp>
+#include <vespa/vespalib/objects/deserializer.hpp>
+
+namespace search::expression {
+
+using vespalib::Deserializer;
+using vespalib::Serializer;
+
+IMPLEMENT_IDENTIFIABLE_NS2(search, expression, AndPredicateNode, MultiArgPredicateNode);
+
+AndPredicateNode::AndPredicateNode() noexcept = default;
+
+AndPredicateNode::~AndPredicateNode() = default;
+
+AndPredicateNode::AndPredicateNode(const std::vector<FilterPredicateNode>& input)
+  : MultiArgPredicateNode(input)
+{
+}
+
+Serializer& AndPredicateNode::onSerialize(Serializer& os) const {
+    return os << args();
+}
+
+Deserializer& AndPredicateNode::onDeserialize(Deserializer& is) {
+    is >> args();
+    return is;
+}
```

Comment:
`onSerialize` and `onDeserialize` is currently duplicated for `AndPredicate` and `OrPredicate`. Since they are both handled by `MultiArgPredicateNode` in Java, it makes sense for this to be the case here as well.

Also, since `IdentifiablePtr` is serialized as an _optional_ object, it seems to me that we either have to detect `nullptr`-ness of deserialized objects here and fail the deserialization with an appropriate exception, or we have to skip `nullptr` objects during traversal. I prefer the former since I don't see any valid semantics for missing elements for predicates.

Otherwise we'll send the search core to Narnia when dereferencing missing objects during calls to `allow`.